src/rpc/tensorpipe/tp_communicator.cc

```diff
@@ -0,0 +1,154 @@
+/*!
+ *  Copyright (c) 2019 by Contributors
+ * \file tp_communicator.cc
+ * \brief Tensorpipe Communicator for DGL distributed training.
+ */
+
+#include "tp_communicator.h"
+
+#include <time.h>
+#include <unistd.h>
+
+#include <future>
+#include <memory>
+#include <utility>
+
+#include "../rpc.h"
+
+namespace dgl {
+namespace rpc {
+
+using namespace tensorpipe;
+
+void TPSender::AddReceiver(const std::string& addr, int recv_id) {
+  receiver_addrs_[recv_id] = addr;
+}
+
+bool TPSender::Connect() {
+  for (auto kv : receiver_addrs_) {
+    std::shared_ptr<Pipe> pipe;
+    for (;;) {
```

Comment:
endless retrying if no connection could be built? Is it on-purpose? or could we skip unavailable ones after retrying several times?

Reply:
Current distributed design requires all clients connecting to server at the beginning. Therefore if some are unavailable, the whole program just steps into the failure stage

Reply:
so some logic exist outside to detect and step into failure stage if some connections are unavailable? timeout?

Reply:
Currently the answer is no.